### ISSUE TITLE
Unable to use PROXY_ADDRESS

### ISSUE THREAD

**Issue by cyrilgdn** [jippi/hashi-ui#200]
Hello,

Maybe I misunderstood the good way to do but I tried to use PROXY_ADDRESS and was unable to make it worked.

If I run 
```
docker run --rm -e NOMAD_ENABLE=true -e NOMAD_ADDR=http://nomad.service.consul:4646 -p 3000:3000 jippi/hashi-ui:v0.13.1
```
I can access to Nomad UI on http://localhost:3000/ and it works perfectly.

But if I try to access on http://localhost:3000/test with
```
docker run --rm -e PROXY_ADDRESS="localhost:3000/test" -e NOMAD_ENABLE=true -e NOMAD_ADDR=http://nomad.service.consul:4646 -p 3000:3000 jippi/hashi-ui:v0.13.1

(1) 2017/02/07 15:40:52.408171 {"level":"info","msg":"application created","context":{"app":"hashi-ui","enabled":false,"version":"1.5.0"}}
15:40:52.408 main.go:76 ▶ INFO  -----------------------------------------------------------------------------
15:40:52.408 main.go:77 ▶ INFO  |                             NOMAD UI                                      |
15:40:52.408 main.go:78 ▶ INFO  -----------------------------------------------------------------------------
15:40:52.408 main.go:79 ▶ INFO  | listen-address  	: http://0.0.0.0:3000                                |
15:40:52.408 main.go:80 ▶ INFO  | proxy-address   	: localhost:3000/test                                |
15:40:52.408 main.go:81 ▶ INFO  | log-level       	: info                                               |
15:40:52.408 main.go:89 ▶ INFO  | nomad-enable     	: true                                               |
15:40:52.408 main.go:93 ▶ INFO  | nomad-read-only       : No (Hashi-UI can change Nomad state)               |
15:40:52.408 main.go:95 ▶ INFO  | nomad-address         : http://nomad.service.consul:4646                   |
15:40:52.408 main.go:96 ▶ INFO  | nomad-ca-cert         :                                                    |
15:40:52.408 main.go:97 ▶ INFO  | nomad-client-cert     :                                                    |
15:40:52.408 main.go:98 ▶ INFO  | nomad-client-key      :                                                    |
15:40:52.408 main.go:99 ▶ INFO  | nomad-skip-verify   	: false                                              |
15:40:52.408 main.go:103 ▶ INFO  | nomad-skip-verify     : No                                                 |
15:40:52.408 main.go:107 ▶ INFO  | consul-enable     	: false                                              |
15:40:52.409 main.go:111 ▶ INFO  | consul-read-only     : No (Hashi-UI can change Consul state)              |
15:40:52.409 main.go:113 ▶ INFO  | consul-address       : 127.0.0.1:8500                                     |
15:40:52.409 main.go:115 ▶ INFO  -----------------------------------------------------------------------------
15:40:52.409 main.go:116 ▶ INFO  
15:40:52.418 nomad.go:24 ▶ INFO  Starting handlers for region: eu-central
15:40:52.418 nomad.go:43 ▶ INFO    -> Connecting to nomad
15:40:52.418 nomad.go:52 ▶ INFO    -> Starting resource watchers
15:40:52.418 main.go:130 ▶ INFO  Nomad client successfully initialized
15:40:52.419 main.go:211 ▶ INFO  Listening ...


```
The page diplay this error:
![nomad-ui](https://cloud.githubusercontent.com/assets/4532947/22697327/655acf94-ed51-11e6-8fff-122eeccde32f.png)

Did I missed something ? 

Thanks in advance.

**Comment by cyrilgdn** [comment]
By the way in the source code I saw a NOMAD_PROXY_ADDRESS, try both but don't know the goal of this second one.

**Comment by jippi** [comment]
Hi

In dev mode only `/nomad` and `/consul` work - `/` and other endpoints will result in a failure

**Comment by cyrilgdn** [comment]
Thanks for your quick answer. 

I encountered some troubles behind a load balancer which strip the prefix too, that's why I made this simple test. But honestly I'm not sure this was the same error...

I will retry my test with the load balancer and will come back to you if I still have an error.

**Comment by cyrilgdn** [comment]
Hi @jippi ,
I tested with a simple Ngninx Proxy:
```
server {
    listen 80;
    server_name localhost;
    location /test {
        # Tried with and without this section, which seems to allow websocket proxying?!
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_http_version 1.1;
        # 

        proxy_pass http://localhost:3000;
    }
}
```
With the docker command:
```
docker run --rm -e NOMAD_PROXY_ADDRESS="localhost/test" -e NOMAD_ENABLE=true -e NOMAD_ADDR=http://nomad.service.consul:4646 -p 3000:3000 jippi/hashi-ui:v0.13.1
```
and I have the same error.

**Comment by burdandrei** [comment]
We're truing to run Hashi-UI behing traefik, and it's not working =(

**Comment by jippi** [comment]
I've been home sick all day today, so didn't get a chance to look into this... will try to set it up myself and test wether it works or not - could be a bug :)

**Comment by cyrilgdn** [comment]
I tried with Traefik too (which is actually the loadbalancer I use, Nginx was only to put a simple example for testing)

@jippi I hope you get better soon!

**Comment by pgoultiaev** [comment]
+1 Also tried using traefik and ran into the same issue.

**Comment by burdandrei** [comment]
i think this is fixed now https://github.com/containous/traefik/issues/1059

**Comment by burdandrei** [comment]
Update here. it was not working with AWS ELB.
AWS ELB doesn't suuport websockets on layer 7 (HTTP), when i switched it to TCP Listener it started to work with Traefik like a magic!

**Comment by jippi** [comment]
oh, nice!

@burdandrei is it something you can contribute in README that AWS ELB need some TLC to work out of the box? :)

**Comment by burdandrei** [comment]
https://github.com/jippi/hashi-ui/pull/203

**Comment by multani** [comment]
Actually, the original problem wasn't talking about AWS, nor ELB nor any kind of proxy in front of the UI at first :)

AFAIK, the following command should serve the UI on http://localhost:3000/nomad/ but it doesn't fully work: `docker run --rm -e PROXY_ADDRESS="localhost:3000/nomad/" -e NOMAD_ENABLE=true -e NOMAD_ADDR=http://nomad.service.consul:4646 -p 3000:3000 jippi/hashi-ui:v0.13.1`.

We get "Error: Unable to connect to the backend..." errors after a couple of seconds opening the http://localhost:3000/nomad/ and the browser network console seems to show that the query to `/nomad/` asking to upgrade to a Websocket connection has some problem, although it returns a 200 error code.

The browser console logs show the following (the opened HTTP request's response is empty):
![screenshot from 2017-02-15 12-11-30](https://cloud.githubusercontent.com/assets/65311/22971975/f39aebbe-f377-11e6-8b18-f65a0b67c322.png)

Is there something we are missing from Docker run command above?

**Comment by jippi** [comment]
@vancluever did i break anything in your proxy support? seem to be problem in either how i've documented it or how it's expected to work :) Your feedback would be lovely

**Comment by vancluever** [comment]
Hey @jippi and all! I'll take a look tonight and see if I can re-produce - more info as I have it!

**Comment by vancluever** [comment]
OK, so I think something is going on with the fact that it seems like the URL routes have changed significantly. Just reviewing (have to re-factor the frontend I made to test this) and will update soon!

**Comment by vancluever** [comment]
Alright, gotten as far as I can spend on it for tonight, but have found a few things:

 * Looks like the single-function URI redirects (ie: when the UI is functioning either in Consul-only or Nomad-only mode) are not including the proxy URI. I've fixed this, but now have run into issues with connecting to the websocket, possibly due to how I fixed it, but more on that below.
 * Right now, we don't include the protocol scheme (ie: `http` or `https`) when asking users to specify the proxy address. I think we may need to change that, since in order to properly pass the correct redirect, we will need to parse the URI from the address, and the easiest way is `url.Parse`, which fails on a bareword URL like `example.com/nomad`, for example.

In any case, this should serve as a good starting point. I'll try and put in some time over the next few days to get this fixed so that it works with the current state of things. May need some other eyes when it's all complete.

**Comment by jippi** [comment]
Thanks for your debugging and findings so far @vancluever ! :)

**Comment by multani** [comment]
Thanks Chris for having a look!



On Thu, Feb 16, 2017, at 04:18, Chris Marchesi wrote:
>  * Right now, we don't include the protocol scheme (ie: http or https)
>    when asking users to specify the proxy address. I think we may need
>    to change that, since in order to properly pass the correct
>    redirect, we will need to parse the URI from the address, and the
>    easiest way is url.Parse, which fails on a bareword URL like
>    example.com/nomad, for example.


If I may add some comment here: it would be *really* awesome if it could
be possible to NOT have to specify the protocol AND the hostname.


Ideally, we would like to deploy this kind of application under a
dedicated prefix, such as "/nomad/" in this case, and be able to reuse
this same job into different environments, without needing to change
either the protocol or the hostname because it's different from one
place to another.


A particular annoying example of an application requiring this kind of
setting is phpMyAdmin which requires the PmaAbsoluteURI parameter to be
set to the complete URI on which this instance of phpMyAdmin will be
reachable (so something like http://my.intranet/phpmyadmin/), which
means you need to spawn a new instance with another configuration if you
want to reach this application from, say
https://extranet.mycompany.com/phpmyadmin/.


Thanks again for the initial patch and the upcoming fixes!

**Comment by jippi** [comment]
I've made a PR that seem to fix the issues, please see #204 - closing this ticket and moving discussion to the PR :)
